src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,385 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    internal sealed partial class CSharpUseRecursivePatternsCodeRefactoringProvider
+    {
+        private abstract class AnalyzedNode : Visitor<AnalyzedNode>
```

Comment:
before continuing, can you give an intuition abotu what's goign on in this design?

Reply:
I'll do a comment pass in the code .. 

nodes themselves are a Visitor which yields a possibly reduced node. this is to handle various combinations without type checks and dealing with symmetrical unifications.

Reply:
> nodes themselves are a Visitor

This itself warrants some good explanations.  This is a new way of approaching things that doesn't seem to have parallels to anything done in our refactorings before.  I don't have a good intution of what it means to have a tree of *visitors*.  So going in depth on what this all means would be great :)

Reply:
Another way to do this would be a big pattern switch with two inputs which checks in any other,
```cs
switch (left, right) {
  case (NodeA, NodeB) 
  case (NodeB, NodeA):
    // ...
```
This corresponds to one of static Visit methods on the base node. Since we don't have that luxury yet, this is the most efficient technique I could come up with.